README
***********************************************************************

             FONLL            version 1.3.3 (April 2014)

 Program to calculate heavy quark transverse momentum and
 rapidity distributions in hadron-hadron and photon-hadron
 collisions, matching Fixed Order next-to-leading order terms
 and Next-to-Leading-Log large-p_T resummation.

 by Matteo Cacciari, Stefano Frixione, Paolo Nason

 Citation for this work:
		 
1)  M. Cacciari, M. Greco and P. Nason,
       "The p_T Spectrum in Heavy Flavor Hadroproduction",
       JHEP 9805 (1998) 007 [hep-ph/9803400]
2)  M. Cacciari, S. Frixione and P. Nason,
       "The p_T Spectrum in Heavy Flavor Photoproduction",
       JHEP 0103 (2001) 006 [hep-ph/0102134]

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\cite{Cacciari:1998it}
\bibitem{Cacciari:1998it}
M.~Cacciari, M.~Greco and P.~Nason,
%``The p(T) spectrum in heavy-flavour hadroproduction,''
JHEP {\bf 9805} (1998) 007
[arXiv:hep-ph/9803400].
%%CITATION = HEP-PH 9803400;%%
%\cite{Cacciari:2001td}
\bibitem{Cacciari:2001td}
M.~Cacciari, S.~Frixione and P.~Nason,
%``The p(T) spectrum in heavy-flavor photoproduction,''
JHEP {\bf 0103} (2001) 006
[arXiv:hep-ph/0102134].
%%CITATION = HEP-PH 0102134;%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
				 

  Built upon the following papers/codes:
			 
  S. Frixione and P. Nason,
       "Phenomenological study of charm photoproduction at HERA",
       JHEP 0203 (2002) 053 [hep-ph/0201281]
  M. Cacciari, S. Frixione and P. Nason,
       "The p_T Spectrum in Heavy Flavor Photoproduction",
       JHEP 0103 (2001) 006 [hep-ph/0102134]
  M. Cacciari, M. Greco and P. Nason,
       "The p_T Spectrum in Heavy Flavor Hadroproduction",
       JHEP 9805 (1998) 007 [hep-ph/9803400]
  M. Cacciari and M. Greco,
       "Charm Photoproduction via Fragmentation",
       Z.Phys. C69 (1996) 459 [hep-ph/9505419] 
  M. Cacciari and M. Greco,			 
       "Large p_T Hadroproduction of Heavy Quarks",
       Nucl.Phys. B421 (1994) 530 [hep-ph/9311260]
  B. Mele and P. Nason, 
       "The Fragmentation Function for Heavy Quarks in QCD",
       Nucl.Phys. B361 (1991) 626
  P. Nason, S. Dawson and R.K. Ellis,
       "The One-Particle Inclusive Differential Cross-Section
	for Heavy Quark Production in Hadronic Collisions",
       Nucl.Phys. B327 (1989) 49
  F. Aversa, P. Chiappetta, M. Greco and J.-Ph. Guillet,
       "QCD Corrections to Parton-Parton Scattering Processes",
       Nucl.Phys. B327 (1989) 105
  R.K. Ellis and P. Nason,
       "QCD Radiative Corrections to the Photoproduction
	of Heavy Quarks",
       Nucl.Phys. B312 (1989) 551
  P. Aurenche, R. Baier, A. Douiri, M. Fontannaz and D. Schiff,
       "Scheme Invariant Higher Order QCD Predictions for
	Large p_T Photoproduction Reactions",
       Nucl.Phys. B286 (1987) 553

**********************************************************************

The installation instructions are given in the file INSTALL. See that
file also for instructions about running a test of the installed code.

The FONLL code produces a SINGLE point of the double-differential cross
section dsigma/dpT^2/dy, and outputs it (in picobarn) in a file .out
with format pt,y,dsigma

A file .outlog is created  containing all the relevant inputs of a run,
and the various components of the final result. All the results of runs
made with the same name tag are appended to this file.

If integrated cross sections are needed, it is up to the user to collect
these differential results and to integrate them.

A grid can be produced either by running the whole code several times,
or by building a Fortran code which loops over the subroutine fonll0
(see the file FONLL/main/fonll0.f, and how it is called from
FONLL/main/fonll.f, which is the main program of the FONLL code).

When run, the program describes the inputs it expects. Sample input files
are as follows:

1) bottom production at the Tevatron (run 2) using the hvqpdfpho.f 
   library (cteq6m in this case):
**********************************************************************
b-Tevatron        ! tag for output files
  1  980. 0 0 131 ! beam1: type, ener.,  nptype, ngroup, nset
 -1  980. 0 0 131 ! beam2: type, ener.,  nptype, ngroup, nset
  5               ! heavy quark mass
  -1              ! Lambda_5, <=0 for default
  1   1           ! ren. and fact. scale factors
  30. 0.          ! pt,y_lab
  1               ! icalctype
**********************************************************************
This should return the file b-Tevatron.out containing three numbers
per line: pt, y and dsigma/dpT^2/dy (in picobarn)

In b-Tevatron.outlog a move verbose output is written,
reporting all input parameters and several partial results:
hdmv=hadronic massive O(as^3) result (FO in ref.1, FO_{hdr} in ref. 2)
hdml=as above, without mass terms (FOM0 in ref.1, FOM0_{hdr} in ref. 2)
phmv=in case of photoproduction or electroproduction, O(as^2 aem) cross section
     (FO_{pnt} in ref. 2).
phml=as above, without mass terms (FOM0_{pnt} in ref. 2).
hdrs=hadroproduction resummed result (RS in ref.1, RS_{hdr} in ref. 2)
phrs=pointlike photon resummed result (RS_{pnt} in ref.2)

In the case of hadroproduction, the photonic components of the cross
section are reported to be zero.

2) charm electroproduction at HERA (H1 ETAG33 setup):
**********************************************************************
c-HERA          ! tag for output files
 1  920. 0 0 81 ! beam1: type, ener.,  nptype, ngroup, nset
 5  27.5 0 0 42 ! beam2: type, ener.,  nptype, ngroup, nset
  1.5           ! heavy quark mass
  -1            ! Lambda_5, <=0 for default
  1    1        ! ren. and fact. scale factors
  10. 1.        ! pt,y_lab
 1              ! icalctype
 1              ! WW type  
.1              ! max. sqrt(Q^2) (GeV)
.29  .62        ! zmin and zmax in the WW integration
**********************************************************************
Note the choice of a PHOTON PDF set (AFG in this case) for doing
electroproduction. The FONLL code handles the convolution with the
Weizsaecker-Williams function.
Results will be in c-HERA.out and c-HERA.outlog


The program incorporates three different interfaces to Parton 
Distribution Function libraries, which can be selected by
building the proper executable:

- A proprietary one is given by the file hvqpdfpho.f.
  Data files are collected in FONLL/pdfdata/. The numbering scheme can be 
  found in the file FONLL/common/hvqpdfpho.f

- The interface to the (old and not anymore maintained) PDFLIB library
  is provided by the file common/jetpdflib.f
  When linking the code to the PDFLIB package from the CERN libraries,
  three numbers have to be provided in order to determine the PDF set,
  rather than only one.
  NB: our interface to PDFLIB compensates for a bug which multiplies by
  two the outputs of the AFG photon PDF set. As of version 8.04 of PDFLIB
  this bug is still present. If removed, one should then update the
  interface.

- Finally, the program also provides an interface to the modern PDF
  library LHAPDF via the file common/lhapdfif.f 
  In this case a single number, in the place of the one used by 
  hvqpdfpho.f, has to be provided to indentify the pdf set (for
  instance "0 0 10050" will choose cteq6m)
  The AFG bug is present here too, and has finally been corrected in 
  LHAPDF v5.8.5. 
     - FONLL <= v1.3.2 was always fixing the bug itself, so it is fine
       to use with any version of LHAPDF < 5.8.5. 
     - FONLL >= v1.3.3 checks for the LHAPDF version used before deciding 
       whether to fix the bug or not.
     *** It is therefore WRONG to use FONLL <= v1.3.2 with LHAPDF >= v5.8.5 
         for photoproduction or electroproduction ***

Nuclear PDFs, as implemented either in EKS98 or EPS09, can also be used.
 * For EKS, the beam type should be set to -A, where A is the atomic number
   (for instance, -208 will give a Pb beam)
 * For EPS, where multiple error sets exist, the beam type should to
   -(A*1000+pset), where pset is the number of the error set, running from
   1 (central set) to 31 (e.g. -208002)
In both cases the appropriate files containing the PDF sets should be copied 
or linked to the working directory.


The program produces a number of internal files; they all have the form
*.tmp. They can be safely removed.

The other files produced are
tag.out
tag.outlog
tag-fonll.log
where 'tag' is the prefix string entered for the run.
The first two files contain the output of the run. The .log file
contains the input used for the run. Suitably renamed, it can be used
as input file for the run (using redirection).


The FONLL program does not provide for the convolution with a non-perturbative
fragmentation function. The typical time to produce a point in pt and y
is of the order of a minute; it is thus convenient to generate a grid
of cross section values, to be interpolated later in order to perform
the convolution with a non-perturbative fragmentation function.
It is up to the user to make sure that the grid he is using
is sufficient to yield good interpolation. We do however provide
examples of how to set up a grid and interpolate it.
Two different methods are illustrated in the directories misc1 and misc2.
These are provided as a basic set of tools to be modified according to
the user's needs in order to obtain a cross section which includes
a convolution with a fragmentation function, and eventually also
an integrated cross section with given cuts.



ChangeLog
*** Version 1.3.3 ***

April 10, 2014

        Eliminated the last (hopefully) vestige of output to an auxiliary
	but useless file (phres.tmp) in phresummed/photon_wps8.f

April 9, 2014

        Inserted an LHAPDF version number test in common/lhapdfif.f,
	so as to check whether the hack to fix the AFG photon PDF
        set (n. 363) should still be used, since the bug has been fixed 
	in LHAPDF since release 5.8.5 -02/02/2011

        This means that FONLL >= v1.3.3 can only be used with an LHAPDF version
	that supports the call to getlhapdfversion(). This function seems
	to be present at least as early as LHAPDF v5.5.0, which dates from 2008


*** Version 1.3.2 ***

July 25, 2012

	Replaced "double precision" with "real*8" in common/hvqpdfpho.f,
	and made explicit the return type of ctq61pdf() in cteq61() for
	gfortran compatibility

September 30, 2010

        Fixed a compatibility problem with gfortran 4.4 (toend() produces
        and infinite loop. Replaced with "access='append'" in the open
        statement) in main/fonll.f. This seems to work with g77, g95 and
        gfortran 4.4 on Darwin

        Also, eliminated useless output to phres.tmp in 
        phresummed/photon_wps8.f

May 20, 2010
   
        Added file eps09.f in common (and to the Makefile in main)
	for providing the new EPS09 nuclear PDFs. Modified fonllpdf.f
	accordingly
	

January 29, 2010

        Eliminated writing of hdms-log.tmp, unit 66, in hdmassive/hdms.f
        to try to get rid of pointlike random crashes when multiple
        instances of the code run together in the same directory 

        Same for phmassive/phms.f

        This seems to solve most of the problems, but not all of them.

December 8, 2009

        Compactified long equation in hdresumment/struv_plus_avgo.f
        because the Intel ifort compiler does not accept more than 511
        continuation lines.

        COMPILATION WITH ifort MUST INCLUDE THE -save FLAG

November 11, 2008

        Fixed a bug in pdfpar() in common/lhapdfif.f which was
        preventing the use of LHAPDF with electrons.
        In the same routine added a check preventing repeated
        printout of the alpha_s value (happened when using photons)

July 9 and 11, 2008

        Added save statements to alfas() in common/hvqpdfpho.f,  
	alfas_p() and air() in common/fonfra_p.f, phdfpy() in 
        phmassive/phms.f, dfpy() in hdmassive/hdms.f, fonpdf() and 
        afgpdf() in common/hvqpdfpho.f  to make them compatible with 
        gfortran and g95. "save" added to all routines in hvqpdfpho.f
        containing an "if(init..." block.

        NB: "save" statemants are probably missing in other routines

        NB2: OPTIMIZATION (-O) SHOULD NOT BE USED WITH GFORTRAN OR G95
             UNLESS A -fno-automatic STATEMENT IS ALSO INCLUDED

*** Version 1.3.1 ***

Februray 1, 2008
        common/lhapdfif.f: the lambda5 value is now calculated
        from the alpha_s(mz) one. This avoids problems with sets
        which return 0 (like cteq6.5 and, more importantly, with
        MRST sets which return a slightly different lambda5, leading
        to an incorrect value for alpha_s
    

*** Version 1.3.0 ***

December 17, 2007
        Eliminated output to unit 32, hdrsout.tmp, in hdresummed/main_p.f,
        as it was essentially useless

        Patched output format (i1->i2) for isuda variable 
        and for pdf set in hdresummed/input.f

        Suppressed output to unit 15, hdres.tmp, from hdresummed/input.f 
        and hdresummed/main_p.f (can be reinstated by setting ioutput=15
        in hdresemmed/input.f)

        Shifted the wmax limit from 1 to 1-1d-16 in 
        dplus() in hdresummed/main_p.f to avoid a 1/1-w division by zero
        error in the w-1 limit

December 9, 2007
        Fixed LIBSLHA variable in main/Makefile so that it uses
        lhapdf-config from the LHAPDF installation. It must, of
        course, be correctly included in PATH beforehand

        LHAPDF does not return correctly lambda_QCD for CTEQ65.
        Value added by hand in common/lhapdfif.f

October 22, 2007
	IMPORTANT CHANGE: Added interface to LHAPDF, file common/lhapdfif.f

	modified fonllpdf.f : removed mlmpdf8 (moved to hvqpdfpho.f)
	                      jetpdflib now calls directly mlmpdf8

	renamed param to hdrsparam in hdresummed/input.f, main_p.f
	
September 3, 2007
        IMPORTANT CHANGE:
        Similarly to the massless calculation, the resummed one also suffers
        from numerical precision problems in certain limits, for 
        instance at large rapidity.
        The combination STRUV*+AVGO in the function fresc() in 
        hdresummed/main_p.f has been reimplemented in the function
        struv_plus_avgo in hdresummed/struv_plus_avgo.f by simplifying
        the proper matrix element contribution via Mathematica.
        This (together with the similar simplification described below) 
        should fix most of the numerical precision problems.
        
Aug. 30, 2007
        Fixed the wrapper to cteq6.1 in common/hvqpdfpho.f
        Bug was instead leading to the naming convention of cteq6 for
        data tables

Aug. 19, 2007
        IMPORTANT CHANGE:
	The function hqhpggm0 in hdmassive/rozlimgg.f lost precision
        in the large rapidity limit. This should now be fixed
        by a rewriting of part of the formula, through a simplification
        obtained with Mathematica.
	
	Added the option icalctype=4, to calculate massive and massless
	only (i.e. no resummed)

        Added to the function dfpy() in hdmassive/hdms.f the option
	to perform the integrals using VEGAS. Actually not used
	(hardcoded), because the precision looks good enough with the
	old (faster) option
		
          
Jul. 6, 2007
        Passed the input epsint to dovegas, so that the precision
        of the integral can be set in fonll0.f (default is 0.001).
        Also set the precision of the xes1 integral to epsint*10
        in hdresummed/main_p.f.

main/fonll.f
************************************************************************
*                                                                      *
*                    FONLL             version 1.3.3 (Apr 2014)        *
*                                                                      *
*      Program to calculate heavy quark transverse momentum and        *
*      rapidity distributions in hadron-hadron and photon-hadron       *
*      collisions, matching Fixed Order next-to-leading order terms    *
*      and Next-to-Leading-Log large-p_T resummation.                  *
*                                                                      *
*      by Matteo Cacciari, Stefano Frixione, Paolo Nason               *
*                                                                      *
*      Citation for this work:                                         *
*                                                                      *
*      M. Cacciari, S. Frixione and P. Nason,                          *
*           "The p_T Spectrum in Heavy Flavor Photoproduction",        *
*           JHEP 0103 (2001) 006 [hep-ph/0102134]                      *
*      M. Cacciari, M. Greco and P. Nason,                             *
*           "The p_T Spectrum in Heavy Flavor Hadroproduction",        *
*           JHEP 9805 (1998) 007 [hep-ph/9803400]                      *
*                                                                      *
*                                                                      *
*      Built upon the following papers/codes:                          *
*                                                                      *
*      S. Frixione and P. Nason,                                       *
*           "Phenomenological study of charm photoproduction at HERA", *
*           JHEP 0203 (2002) 053 [hep-ph/0201281]                      *
*      M. Cacciari, S. Frixione and P. Nason,                          *
*           "The p_T Spectrum in Heavy Flavor Photoproduction",        *
*           JHEP 0103 (2001) 006 [hep-ph/0102134]                      *
*      M. Cacciari, M. Greco and P. Nason,                             *
*           "The p_T Spectrum in Heavy Flavor Hadroproduction",        *
*           JHEP 9805 (1998) 007 [hep-ph/9803400]                      *
*      M. Cacciari and M. Greco,                                       *
*           "Charm Photoproduction via Fragmentation",                 *
*           Z.Phys. C69 (1996) 459 [hep-ph/9505419]                    *
*      M. Cacciari and M. Greco,                                       *
*           "Large p_T Hadroproduction of Heavy Quarks",               *
*           Nucl.Phys. B421 (1994) 530 [hep-ph/9311260]                *
*      B. Mele and P. Nason,                                           *
*           "The Fragmentation Function for Heavy Quarks in QCD",      *
*           Nucl.Phys. B361 (1991) 626                                 *
*      P. Nason, S. Dawson and R.K. Ellis,                             *
*           "The One-Particle Inclusive Differential Cross-Section     *
*            for Heavy Quark Production in Hadronic Collisions",       *
*           Nucl.Phys. B327 (1989) 49                                  *
*      F. Aversa, P. Chiappetta, M. Greco and J.-Ph. Guillet,          *
*           "QCD Corrections to Parton-Parton Scattering Processes",   *
*           Nucl.Phys. B327 (1989) 105                                 *
*      R.K. Ellis and P. Nason,                                        *
*           "QCD Radiative Corrections to the Photoproduction          *
*            of Heavy Quarks",                                         *
*           Nucl.Phys. B312 (1989) 551                                 *
*      P. Aurenche, R. Baier, A. Douiri, M. Fontannaz and D. Schiff,   *
*           "Scheme Invariant Higher Order QCD Predictions for         *
*            Large p_T Photoproduction Reactions",                     *
*           Nucl.Phys. B286 (1987) 553                                 *
*                                                                      *
************************************************************************
c Driving program for FONLL calculations
      program fonll
      implicit none
c verion number string
      include 'version.h'
c prefix for generated files
      character * 30 prefix
c jprefix is last non-blank in prefix
      integer jprefix
      integer ibeam1,ibeam2,nptype1,nptype2,ngroup1,ngroup2,nset1,nset2
     # ,icalctype
      real * 8 ener1,ener2,xmq,xlam,ffact,fren,pt,ylab
      real * 8 hdmassive,hdmassless,hdresummed,hdreserr,sigmafonll,
     #         phmassive,phmassless,phresummed,phreserr
      real*8 ratio,reserr
c ww common
      integer itypeww
      real * 8 xmuww,thcww,zminww,zmaxww,elphen
      common/fonllww/xmuww,thcww,zminww,zmaxww,elphen,itypeww
c Unit used for temporary files. Keep closed
c before calling fonll0
      integer itm
      data itm/57/
      write(*,*) ' enter prefix for this run'
      read(*,*) prefix
      do jprefix=len(prefix),1,-1
         if(prefix(jprefix:jprefix).ne.' ') goto 11
      enddo
 11   continue
      open(unit=itm,file=prefix(1:jprefix)//'fonll.log')
      write(*,*) ' pdf type is described by 3 numbers'
      write(*,*) ' pdflib uses all 3; other may use less'
      write(*,*) ' All energy values are in GeV'
c
      write(*,*) ' Beam type are:'
      write(*,*) ' -1 for p_bar, 1 for p, 0 for (n+p)/2, 2 for n,'
      write(*,*) ' 3 for pi^+, -3 for pi^-'
      write(*,*) ' 4 for photon, 5 for electron'
c
      write(*,*) ' beam1 has positive rapidity'
c
      write(*,*)
      write(*,*) 'enter beam 1 type, energy, 3 pdf numbers'
      read(*,*)  ibeam1,ener1, nptype1, ngroup1, nset1
      write(itm,*) ibeam1,ener1, nptype1, ngroup1, nset1,
     # ' ! beam1: type, ener.,  nptype, ngroup, nset'
      write(*,*)
      write(*,*) 'enter beam 2 type, energy, 3 pdf numbers'
      read(*,*)  ibeam2,ener2, nptype2, ngroup2, nset2
      write(itm,*) ibeam2,ener2, nptype2, ngroup2, nset2,
     # ' ! beam2: type, ener.,  nptype, ngroup, nset'
c
      write(*,*)
      write(*,*) ' heavy quark mass'
      read(*,*) xmq
      write(itm,*)xmq,' ! heavy quark mass'
      write(*,*)
      write(*,*) 'LambdaQCD_5 in GeV, (<=0 for default'
      read(*,*) xlam
      write(itm,*) xlam, ' ! Lambda_5, <=0 for default'
c
      write(*,*)
      write(*,*) 'fact. and ren. scale factors'
      read(*,*) ffact,fren
      write(itm,*) ffact,fren,' !  fact. and ren. scale factors'
c
      write(*,*)
      write(*,*) 'enter pt,y_lab'
      read(*,*) pt,ylab
      write(itm,*) pt,ylab, ' ! pt,y_lab'
c
      write(*,*)
      write(*,*) ' enter 1 for FONLL, 2 for Fixed Order, 3 for '//
     # 'resummed in the massless limit (no power suppressed mass terms)'
      read(*,*) icalctype
      write(itm,*) icalctype, '   ! icalctype'
      if(ibeam1.eq.5.or.ibeam2.eq.5) then
         write(*,*) ' Weiszaeker-Williams parameters'
         write(*,*)'enter 1 for cut in photon virtuality'
         write(*,*)'      2 to cut on angle'
         write(*,*)'      3 as in 1, only log term'
         write(*,*)'      4 as in 2, only log term'
         read(*,*) itypeww
         write(itm,*) itypeww,'    ! itype ww (1-4)'
         if(itypeww.eq.1.or.itypeww.eq.3)then
            write(*,*)'enter the effective WW scale in GeV (upper limit'
            write(*,*)'of the absolute value of the photon virtuality)'
            read(*,*) xmuww
            write(itm,*) xmuww, '   ! effective ww scale'
         elseif(itypeww.eq.2.or.itypeww.eq.4)then
            write(*,*)'theta_cut'
            read(*,*) thcww
            write(itm,*) thcww, '   ! angular cut for ww electron'
         endif
         write(*,*)
     #  ' enter minimum and maximum energy(photon)/energy(electron)'
         read(*,*) zminww,zmaxww
         write(itm,*) zminww,zmaxww,'! zminww,zmaxww'
      endif
      close(itm)
c End of input
      call fonll0(icalctype,prefix,jprefix,
     #  ibeam1,ener1, nptype1, ngroup1, nset1,
     #  ibeam2,ener2, nptype2, ngroup2, nset2,
     #  xmq,xlam,ffact,fren,pt,ylab,
     #  hdmassive,hdmassless,phmassive,phmassless,
     #  hdresummed,hdreserr,phresummed,phreserr,sigmafonll)
c....used to check higher orders below
      if ( hdmassive+phmassive .gt. 0d0 ) then
          ratio = (hdresummed+phresummed-(hdmassless+phmassless))/
     #   (hdmassive+phmassive)
         reserr = hdreserr + phreserr
      else
          ratio = 0d0
          reserr = 0d0
      endif
      open(11,file=prefix(1:jprefix)//'.outlog',access='append')
c... commented out because a bug (#474382) in gfortran 4.4
c... produces an infinite loop
c wind file to end
c      call toend(11)
c... now using the line above with access='append', which seems to
c... work both in g77 and gfortran 4.4. Note however that the official
c... gfortran syntax is supposed to be
c...      open(11,file=prefix(1:jprefix)//'.outlog',position='append')
c...
      write(11,'(a,a,a)') versionstring
      write(11,111)ibeam1,ener1, nptype1, ngroup1, nset1,
     #  ibeam2,ener2, nptype2, ngroup2, nset2,
     #  xmq,xlam,ffact,fren,pt,ylab,sigmafonll,
     #  hdmassive,hdmassless,phmassive,phmassless,
     #  hdresummed,hdreserr,phresummed,phreserr,ratio
      write(11,*)
      close(11)
 111  format('beam1=',i2,',e1=',d12.6,',pdf1=',i1,',',i1,',',i5,
     #      ',beam2=',i2,',e2=',d12.6,',pdf2=',i1,',',i1,',',i5,/,
     #       'mq=',d12.6,',l5=',d12.6,',ff=',d12.6,',fr=',d12.6,/,
     #       'pt,y,fonll',3(1x,d12.6),/,
     #       'hdmv=',d12.6,',hdml=',d12.6,',phmv=',d12.6,',phml=',d12.6,
     #       / ,'hdrs=',d12.4,'+-',d8.2,',phrs=',d12.6,'+-',d8.2,/,
     #       '(rs-ml)/mv=',d8.2)
      open(11,file=prefix(1:jprefix)//'.out')
c...check if higher orders are very large
c      if(dabs(ratio).gt.5d0)  then
c...check if numerical error on resummed is large
      if(dabs(reserr).gt.0.5*dabs(hdmassive+phmassive))  then
        write(11,'(1x,2(f14.6,1x),e18.9,2x,e15.6,2x,a)') 
     #                 pt,ylab,sigmafonll,reserr,'*'  
c        write(11,'(1x,2(f14.6,1x),e18.9,2x,e15.6,2x,a)') 
c     #                 pt,ylab,hdmassive+phmassive,reserr,'*'  
      else
        write(11,'(1x,2(f14.6,1x),e18.9,2x,e15.6)') 
     #                 pt,ylab,sigmafonll,reserr
      endif      
      close(11)
      write(*,*) ' matched result:', sigmafonll
      write(*,*) ' partial results:'
      write(*,*) ' hadronic mv, ml, rs:',
     # hdmassive,hdmassless,hdresummed
      if(    ibeam1.eq.4.or.ibeam1.eq.5
     #     .or. ibeam2.eq.4.or.ibeam2.eq.5 ) then
         write(*,*) ' pointlike photon mv, ml, rs:',
     #        phmassive, phmassless,phresummed
      endif
      end
